### PR TITLE
[WIP] Add shared docker network for the daemon and services

### LINKED ISSUE #142: Add shared docker network for the daemon and services
In order to be able to use TCP within services

Related to https://github.com/mesg-foundation/service-js-function/issues/1

### PR DESCRIPTION
**Pull request by NicolasMahe**
- Create a shared docker network named "mesg-shared-network"
- Inject `MESG_ENDPOINT_TCP` to services' `ENV` variables.

Close #142

### CODE REVIEW

**Comment by NicolasMahe** [comment]
Some parts of the code are ugly and duplicated because the package `cmd/daemon` and `service` are using common function but they don't share a common `docker` package.
It should be fix in https://github.com/mesg-foundation/core/issues/91

**Comment by NicolasMahe** [comment]
Maybe the service can still have its own network? Let's try before merging!

**Comment by NicolasMahe** [comment]
Will close #91

**Comment by NicolasMahe** [comment]
@antho1404 this PR is ready to be tested!

**Comment by antho1404** [comment]
I put the image as configuration and a script `dev-daemon` and `dev-cli` to start the daemon and cli for dev. With the actual setup if you tag on your computer the image with the same name of the production image `mesg/daemon`, when you run the service, the service will first try to fetch from dockerhub and if it cannot find then look for the image on your disk. This means that even when you build the daemon, when you run it it will still fetch the one from dockerhub so you cannot really test the modification on your docker image. I'm curious to see how you did to do your tests @NicolasMahe I didn't find any other ways to do that

**Comment by antho1404** [comment]
run tests based on the docker image just built for tests also now every time we run tests we need to provide the docker image for the tests. There is some additional setups that I added in the readme

```bash
go test -cover -v ./...	+DAEMON.IMAGE=mesg-daemon-test go test -cover -v ./...
```

If you use Visual code you can add the following settings (Preference > Settings)
```json
"go.testEnvFile": "${workspaceRoot}/testenv"
```

**Comment by antho1404** [comment]
I also added a `test` folder with some integration test. A bug was introduced, the service for the daemon didn't exposed one necessary env variable.

The daemon was looking for the database in a folder that was not shared with the host.

Scenario of the failure:
- `./dev-cli daemon publish ./serviceX` this publish the service in the docker container
- `./dev-cli service list` this read from the disk of the host
The result is that the list doesn't show anything because the service was deployed in a non shared folder because one env variable was not given to the daemon

**Comment by NicolasMahe** [comment]
It's not possible to mount volume on the remote-docker: https://circleci.com/docs/2.0/building-docker-images/#mounting-folders

I try to use the CI in `machine` but I get an other error:
```
#!/bin/bash -eo pipefail
env DAEMON.IMAGE=mesg-daemon-test go test -v -p 1 -coverprofile=coverage.txt ./...
cannot use test profile flag with multiple packages
Exited with code 1
```
maybe it's because of the go version available in the `machine`.

I also add a `ListTasks` and `TasksError` to get the errors of why containers do not start.

I commented the integration test in order for the tests to pass.

I didn't make more test yet and didn't look at the `service test` command.
So it's not ready to merge yet.

File: docker/task.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"context"
+
+	"github.com/docker/docker/api/types/swarm"
+	godocker "github.com/fsouza/go-dockerclient"
+)
+
+// ListTasks returns all docker tasks
+func ListTasks(namespace []string) (tasks []swarm.Task, err error) {
```
Comment (antho1404) [inline docker/task.go]: Is this file useful ? I don't see any use of it

File: config/api.go [hunk 2]
```diff
@@ -35,10 +36,13 @@ func init() {
 	}
 
 	viper.SetDefault(MESGPath, configDir)
+	err = os.MkdirAll(viper.GetString(MESGPath), os.ModePerm)
```
[the remark this answers is not in the document]
Comment (antho1404) [reply config/api.go]: ok this one is valid when the user set the env variable to override the default config directory

**Comment by antho1404** [comment]
I guess this is not what it is supposed to do 

This IP Is not the one of the network, neither the one from the daemon. Maybe it would be helpful that you describes what you want to do to make sure I really understand your code and be able to debug it

screen shot 2018-05-27 at 11 52 42 am

screen shot 2018-05-27 at 11 56 18 am

**Pull request closed without merging**